src/pydvl/influence/array.py

```diff
@@ -0,0 +1,390 @@
+"""
+This module provides classes and utilities for handling large arrays that are chunked and lazily evaluated.
+It includes abstract base classes for converting between tensor types and NumPy arrays, aggregating blocks
+of data, and abstract representations of lazy arrays. Concrete implementations are provided for handling
+chunked lazy arrays (chunked in one resp. two dimensions), with support for efficient storage and retrieval
+using the Zarr library.
+
+Classes:
+    NumpyConverter: Abstract base class for converting between tensor types and NumPy arrays.
+    BlockAggregator: Abstract base class for aggregating blocks of data.
+    ListAggregator: Concrete implementation of BlockAggregator using lists.
+    LazyArray: Abstract base class representing a lazily evaluated array.
+    OneDimChunkedLazyArray: Represents a chunked and lazily evaluated array, where chunking is restricted to the
+        first dimension.
+    TwoDimChunkedLazyArray: Represents a chunked and lazily evaluated array, where chunking is restricted to the
+        two first dimension.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Callable, Generator, Generic, Optional, Tuple
+
+import zarr
+from numpy.typing import NDArray
+
+from .base_influence_model import TensorType
+
+
+class NumpyConverter(Generic[TensorType], ABC):
+    """
+    Base class for converting TensorType objects into numpy arrays and vice versa.
+    """
+
+    @abstractmethod
+    def to_numpy(self, x: TensorType) -> NDArray:
+        """Overwrite this method for converting a TensorType object into a numpy array"""
+
+    @abstractmethod
+    def from_numpy(self, x: NDArray) -> TensorType:
+        """Overwrite this method for converting a numpy array into a TensorType object"""
+
+
+class TensorAggregator(Generic[TensorType], ABC):
+    """
+    Abstract base class for aggregating tensors. Implementations of this class should define
+    methods for aggregating tensors.
+    """
+
+    @abstractmethod
+    def aggregate_from_nested_generators(
```

Comment:
An aggregator is something a user can provide. For example, if the user is only interested in the mean she/he could do
`class TorchMeanAggregator(TensorAggregator):

    def aggregate_from_generator(self, tensor_generator: Generator[torch.Tensor, None, None]):
        return torch.mean(torch.cat([torch.mean(t) for t in tensor_generator]))

...
`
So there is only ever one complete chunk of data in memory. If we call the method flatten, this would be very confusing in my opinion.

Reply:
That makes sense. It's the listaggregator below that's flattening. Understood.

Reply:
Now that I understand this better (hopefully), I see it's just an interface for map-reduce of sorts. There are two steps, correct?

1. Process a structure of nested generators into one generator. (flatten_generators)
2. Process each yielded chunk in some way. (reduce)

Is there a need for two separate functions when it's apparently the same (recursive) process? And could we provide a helper method to produce a generator that "flattens" nested generators? Then users only need implement one method.

So in the end this is just one function `reduce_tensors(reduction_fun, tensors)`, which could actually be `functools.reduce` in some cases, if we provide the generator flattening (although that might be memory inefficient with all the `cat()`s...). I'm just spitballing here.

Reply:
Actually, we don't want to flatten the nested generators (which you could easily do with itertools.chain), because
it resembles the block matrix structure (every inner generator yields the blocks for one row group) and you would loose this representation, if you just flatten the generator. But it might makes sense to split the two functions into different classes, because I never need both methods at the same time. Additionally, the purpose of the class is clearer. But even in this case, it might not be a reduction (list, torch.cat,.....).

Reply:
Understood, thx